File: flink-code-scala-0421/src/main/scala/com/atguigu/day02/ReduceExample.scala
package com.atguigu.day02

import com.atguigu.day02.util.SensorSource
import org.apache.flink.streaming.api.scala._

object ReduceExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream = env.addSource(new SensorSource).filter(r => r.id.equals("sensor_1"))

    stream
      .map(r => (r.id, r.temperature))
      .keyBy(r => r._1)
      .reduce((r1, r2) => {
        (r1._1, r1._2.max(r2._2))
      })
      .print()

    env.execute()
  }
}

File: flink-code-java-0421/src/main/java/com/atguigu/day02/ReduceExample.java
package com.atguigu.day02;

import com.atguigu.day02.util.SensorReading;
import com.atguigu.day02.util.SensorSource;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

public class ReduceExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStream<SensorReading> stream = env.addSource(new SensorSource());

        // 分流，聚合
        // DataStream<SensorReading> => KeyedStream<SensorReading, String> => DataStream<Tuple2<String, Double>>
        stream
                .map(new MapFunction<SensorReading, Tuple2<String, Double>>() {
                    @Override
                    public Tuple2<String, Double> map(SensorReading r) throws Exception {
                        return Tuple2.of(r.id, r.temperature);
                    }
                })
                .keyBy(r -> r.f0)
                .reduce(new ReduceFunction<Tuple2<String, Double>>() {
                    @Override
                    public Tuple2<String, Double> reduce(Tuple2<String, Double> t1, Tuple2<String, Double> t2) throws Exception {
                        if (t1.f1 > t2.f1) {
                            return t1;
                        } else {
                            return t2;
                        }
                    }
                })
                .print();

        env.execute();
    }
}
